Composer/packages/client/src/components/ProjectTree/ProjectTree.tsx

```diff
@@ -197,6 +187,8 @@ export const ProjectTree: React.FC<Props> = ({
   const setPageElement = (name: string, value) => setPageElementState('dialogs', { ...pageElements, [name]: value });
 
   const [filter, setFilter] = useState('');
+  const filterRef = useRef(filter);
```

Comment:
I dont understand what's happening with this ref, what does it represent? and why is it needed?

Reply:
Hi, @hatpick , There is bug in main branch, I fixed it in this PR. The bug is the filter doesn't work when we input some text. The reason is I found the filter text is used in a function, but react will use the stale value anytime
![image](https://user-images.githubusercontent.com/39758135/107379522-f730d000-6b27-11eb-8c21-b498c0ce077b.png)

Reply:
I have created a hook in my local branch for another feature:
```
export const useDebounce = <T>(value: T, delay: number) => {
  const [debouncedValue, setDebouncedValue] = useState<T>(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value]);

  return debouncedValue;
};
```
you can consume it for this exact case like:
```
const [query, setQuery] = React.useState<string | undefined>();
const debouncedQuery = useDebounce<string | undefined>(query, 300);
```

and then you deal with ```debouncedQuery```

wanna give it a try? if it works as expected you can add that hook to share library so others can use it too.

Reply:
Thanks @hatpick , I think I didn't explain the issue very clearly. 
I think there is a wrong usage in < ProjectTree />: Some children components will use the filter message to re-render the UI, but all these components will call a function from parent to filter their items and they don't depends on the filter message change, so the function will only use the stale value. 

So I think I just need to listen to the filter state and re-render these components when filter change. I fixed it in the latest commit and removed the Ref.

I think it's a good idea to share the useDebounce to the library, the filter state in ProjectTree also use a  similar implementation with throttle. Can we do this in a new PR?